OpenRA.Mods.Common/Traits/BotModules/ResourceMapBotModule.cs

```diff
@@ -0,0 +1,313 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[TraitLocation(SystemActors.Player)]
+	public class ResourceMapBotModuleInfo : ConditionalTraitInfo, NotBefore<IResourceLayerInfo>
+	{
+		[Desc("Harvestable and valuable resource types.")]
+		public readonly HashSet<string> ValuableResourceTypes = [];
+
+		[Desc("Tells the AI what types are considered resource creator.")]
+		public readonly HashSet<string> ResourceCreatorTypes = [];
+
+		[Desc($"Actor types that are considered refineries for {nameof(HarvesterTypes)}.")]
+		public readonly HashSet<string> RefineryTypes = [];
+
+		[Desc($"Actor types that are considered harvesters for {nameof(ValuableResourceTypes)}.")]
+		public readonly HashSet<string> HarvesterTypes = [];
+
+		[Desc("Actor types that are considered to be the base building for expansion. Other enemy units will also be recorded",
+			"Defence and production building is suggested")]
+		public readonly HashSet<string> EnemyBaseBuildingTypes = [];
+
+		/* those are options shared by two or more modes */
+		[Desc("Tick in update an indice of resource map.")]
+		public readonly int UpdateResourceMapInverval = 67;
+
+		[Desc("Distance in cells of half indice of the resource map.")]
+		public readonly int ResourceMapStrideRadius = 12;
+
+		public override object Create(ActorInitializer init) { return new ResourceMapBotModule(init.Self, this); }
+	}
+
+	public class ResourceIndice
+	{
+		public int2 IndiceIndexs;
+		public CPos IndiceCenter;
+		public int ResourceCellsCount;
+		public CPos ResourceCellsCenter;
+		public List<CPos> ResourceCreatorLocs;
```

Comment:
```suggestion
		public CPos[] ResourceCreatorLocs = [];
```

There's no need for this to be a list